### PR TITLE
PyBind to bind CarmaClock to Python Module

### PR DESCRIPTION
**Pull request by paulbourelly999**
# PR Details
## Description
This Pull Request adds functionality to create python module bindings for the existing C++ **CARMA Time Library**. It does this using [PyBind11](https://github.com/pybind/pybind11), a light weight, header only library.
Unfortunately do to limitation encounters with downloading and including Python3 for cross-compile targets (armhf, arm32, and arm64) I was only able to get the python module binding to work for native builds (x64). To account for this the, all the python module binding occurs conditionally based on an environment variable `BUILD_ARCHICTECTURE` set in carma-builds for cross compile targets (https://github.com/usdot-fhwa-stol/carma-builds/blob/a77b21967f06c3a8b66c060f8ed4f1e138b7159f/Dockerfile#L25)

> [!IMPORTANT]\
>The changes included here will build python module bindings for all native builds and the generated debian package should include the shared library with these bindings. For cross-compile targets, it will simply skip this process and compile just the C++ library since I ran into many issues trying to correctly include dependencies like python3 for cross-compile targets.
## Related Jira Key
TT-69

## How Has This Been Tested?
test_carma_clock_python_module_binding CMake target

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: carma_clock_py/src/carma_clock_py.cpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#include <pybind11/pybind11.h>
+
+#include <carma_clock.h>
+namespace py = pybind11;
+
+PYBIND11_MODULE(carma_clock_py, m) {
```
Comment (dan-du-car) [inline carma_clock_py/src/carma_clock_py.cpp:6]: Just a suggestion. Maybe rename this to something like "carma_clock_cpp_to_python_adapter"?

File: install_dependencies.sh [hunk 2]
```diff
@@ -1,5 +1,11 @@
...
 set -ex
-
-# TODO
+# export BUILD_ARCHITECTURE=arm64
+# Get ubuntu distribution code name. All STOL APT debian packages are pushed to S3 bucket based on distribution codename.
+echo "BUILD_ARCHITECTURE=${BUILD_ARCHITECTURE}"
+if [[ "${BUILD_ARCHITECTURE}" =~ arm* ]]; then
+    echo "WARNING: Python module support currently only exists for native x64 builds. Skipping Python module binding ... "
+else
+    /opt/carma/scripts/install_dependencies_script.sh python3-dev
```
Comment (dan-du-car) [inline install_dependencies.sh:10]: What happens if this depended install_dependecies_script.sh location changed? How is this scripted downloaded into this predefined path?

File: install_dependencies.sh [hunk 3]
```diff
@@ -1,5 +1,11 @@
 #!/bin/bash
 
 set -ex
-
-# TODO
+# export BUILD_ARCHITECTURE=arm64
```
Comment (dan-du-car) [inline install_dependencies.sh:4]: Where does this export take place if not here?

File: carma_clock_py/python_wrapper_test.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import time
+
+"""Test Case for testing basic CarmaClock functionality in python
+"""
+class TestCarmaClock(unittest.TestCase):
+
+    def testSimClockInitializeException(self):
+        sim_clock = carma_clock_py.CarmaClock(True)
+        with self.assertRaises(ValueError) as e:
+            sim_clock.nowInMilliseconds()
```
Comment (dan-du-car) [inline carma_clock_py/python_wrapper_test.py:13]: Why do we need to call this nowInMilliseconds()?

File: carma_clock_py/python_wrapper_test.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        
+        self.assertIn('Clock is not initialized!',str(e.exception))
+
+    def testRealClockTime(self):
+        clock = carma_clock_py.CarmaClock()
+        self.assertAlmostEqual(time.time()*1000, clock.nowInMilliseconds(),delta=10)
+
+    def testSimClockUpdate(self):
+        sim_clock = carma_clock_py.CarmaClock(True)
```
Comment (dan-du-car) [inline carma_clock_py/python_wrapper_test.py:22]: Why does this initialization success but case 1 failed in initialization?

File: carma_clock_py/README.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+![Dev Container Rebuild Notification](doc/images/DevContainerRebuildNotification.png)
+
+7. If that prompt did not appear, open the **View > Command Palette** and type to locate **"Dev Containers: Rebuild and Reopen in Container"**.
+6. The editor should refresh and display the updated activity in the command line.
+
+![Dev Container Rebuild Log](doc/images/DevContainerRebuildLog.png)
+
+If the container startup was successful, the development environment is up and running and contains the library dependencies and compilation environment. It is possible to proceed with the build.
+
+## Building
```
Comment (dan-du-car) [inline carma_clock_py/README.md:52]: Can we add a sample usage after the build in a python example?

File: carma_clock_py/cmake/dependencies.cmake [hunk 7]
```diff
@@ -0,0 +1,15 @@
+find_package(Python3 REQUIRED COMPONENTS Interpreter Development)
```
Comment (dan-du-car) [inline carma_clock_py/cmake/dependencies.cmake:1]: Do we have a reference where is explained?

File: install_dependencies.sh (hunk 2 above)
Comment (paulbourelly999) [reply install_dependencies.sh:10]: Please look at the documentation in https://github.com/usdot-fhwa-stol/carma-builds . All debian package libraries use carma-builds base images to enable CPack functionality and cross-compiling for ARM debian packages

File: carma_clock_py/python_wrapper_test.py (hunk 5 above)
Comment (paulbourelly999) [reply carma_clock_py/python_wrapper_test.py:22]: Please reference the carma-clock library documentation https://github.com/usdot-fhwa-stol/carma-time-lib/blob/develop/include/carma_clock.h

File: CMakeLists.txt [hunk 8]
```diff
@@ -6,24 +6,51 @@ set(CMAKE_MODULE_PATH "$ENV{CARMA_OPT_DIR}/cmake")
...
+  PUBLIC
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+)
 
-add_library(${PROJECT_NAME} SHARED)
 set_target_properties(${PROJECT_NAME} PROPERTIES
     SOVERSION 1
 )
 
+if (DEFINED ENV{BUILD_ARCHITECTURE})
```
Comment (dan-du-car) [inline CMakeLists.txt:28]: Is this BUILD_ARCHITECTURE only defined with value arm? Can it is defined with value x64?

**Review by dan-du-car — APPROVED** [review]

**Review by kjrush — APPROVED** [review]

**Pull request merged**